XenAdmin/SettingsPanels/VBDEditPage.cs

```diff
@@ -149,16 +150,30 @@ public void UpdateDevicePositions(Session session)
                 {
                     devicePositionComboBox.Items.Clear();
 
-                    devicePositionComboBox.Items.AddRange(devices.ToArray());
+                    var comboBoxItems = vm.Connection.ResolveAll(vm.VBDs).Select(d => new DevicePositionComboBoxItem(d.userdevice, d));
+
+                    foreach (var devicePosition in devices)
+                    {
+                        var comboBoxItemWithVdi = comboBoxItems.FirstOrDefault(ci => ci.VBD.userdevice == devicePosition);
+                        if (comboBoxItemWithVdi != null)
+                        {
+                            devicePositionComboBox.Items.Add(comboBoxItemWithVdi);
+                        }
+                        else
+                        {
+                            devicePositionComboBox.Items.Add(new DevicePositionComboBoxItem(devicePosition));
+                        }
+                    }
 
                     // Make sure the userdevice for the selected VBD is the one selected in the combobox
-                    foreach (String position in devicePositionComboBox.Items)
+                    foreach (DevicePositionComboBoxItem item in devicePositionComboBox.Items)
                     {
-                        if (position != vbd.userdevice)
-                            continue;
+                        if (item.VBD != null && item.VBD.userdevice == vbd.userdevice)
+                        {
+                            devicePositionComboBox.SelectedItem = item;

```

Comment:
I like it as is, because it may iterate the loop again, but only N times where N is the position of the currently selected item. If we moved this code into the previous block it would make that more complex whereas both are trivial now, so much easier to maintain. Re-iteration of this list is cheap anyway, also the layout is suspended while the execution is here)